spec/services/browser_support_spec.rb

```diff
@@ -0,0 +1,99 @@
+require 'rails_helper'
+
+RSpec.describe BrowserSupport do
+  before { BrowserSupport.cache.clear }
+
+  describe '.supported?' do
+    let(:user_agent) do
+      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like ' \
+        'Gecko) Chrome/110.0.0.0 Safari/537.36'
+    end
+
+    subject(:supported) { BrowserSupport.supported?(user_agent) }
+
+    context 'with browser support config file missing' do
+      before do
+        expect(File).to receive(:read).with(Rails.root.join('browsers.json')).
+          and_raise(Errno::ENOENT.new)
+      end
+
+      it { expect(supported).to eq(true) }
+    end
+
+    context 'with invalid support config' do
+      before do
+        expect(File).to receive(:read).with(Rails.root.join('browsers.json')).and_return('invalid')
+      end
+
+      it { expect(supported).to eq(true) }
+    end
+
+    context 'with valid browser support config' do
+      before do
+        allow(BrowserSupport).to receive(:browser_support_config).
+          and_return(['chrome 109', 'ios_saf 14.5-14.8', 'op_mini all'])
+      end
+
+      context 'with nil user agent' do
+        let(:user_agent) { nil }
+
+        it { expect(supported).to eq(false) }
+      end
+
+      context 'with supported user agent' do
+        let(:user_agent) do
+          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like ' \
+            'Gecko) Chrome/110.0.0.0 Safari/537.36'
```

Comment:
is there a way we can maybe comment or hint what rule this is supposed to match to?
```suggestion
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like ' \
            'Gecko) Chrome/110.0.0.0 Safari/537.36' # chrome 110
```

Reply:
> is there a way we can maybe comment or hint what rule this is supposed to match to?

Added user agent comments in e1a1d1958.